feat: add checkbox component

milljoniaer opened the pull request:
![image](https://user-images.githubusercontent.com/78978542/142593292-a17fa1eb-ba76-4fb7-af5f-9620c8dc74e8.png)

### New things:
- new checkbox component
- new storybook story for that component

milljoniaer committed:
feat: introduce checkbox component

milljoniaer committed:
fix: uncheck with only one click
there was a bug, so that unchecking was only possible with two clicks

milljoniaer committed:
feat: add onChange property for access to checked
still some design mistakes

ghost commented:
# [Gatsby Cloud](https://www.gatsbyjs.com) Build Report
**satellytes.com-main**

:tada: Your build was successful! [See the Deploy preview here.](https://satellytescommain21751-featcheckbox.gtsb.io)

## Build Details

[View the build logs here.](https://gatsbyjs.com/dashboard/f215130d-15e3-4176-9252-e13166fe48de/sites/674be368-adf0-4ac5-8fae-f9a4493adc51/builds/6d951ec6-41a2-4367-ada2-dd2a749cea1f/details)

:clock1: Build time: **1m**

## Performance

### Lighthouse report

| Metric | Score |
| ----- | ---- |
| Performance |   :large_orange_diamond: 71   |
| Accessibility |   :green_heart: 97   |
| Best Practices |   :green_heart: 100   |
| SEO |   :large_orange_diamond: 85   |

:link: [View full report](https://build-6d951ec6-41a2-4367-ada2-dd2a749cea1f.gtsb.io/reports/lighthouse/index.html)

milljoniaer committed:
feat: make <input type="checkbox"> fully unvisible

milljoniaer committed:
feat: design update
- Label at correct position
- correct font, sizes and colors
- different color for checkbox story to recognice the
   background of the checkbox

fischeversenker commented on src/components/checkbox/checkbox.tsx:
```diff
@@ -0,0 +1,85 @@
+          </Icon>
+        </StyledCheckbox>
+      </HiddenCheckboxLabel>
+      <LabelText>{props.label}</LabelText>
```
The label is currently not clickable because it is outside of the `label` element. Please move it back inside the `HiddenCheckboxLabel` to improve the UX.

fischeversenker commented on src/components/checkbox/checkbox.stories.tsx:
```diff
@@ -0,0 +1,27 @@
+  title: 'Components/Checkbox',
+  parameters: {
+    layout: 'fullscreen',
+  },
```
This leads to no padding from Storybook. I suggest to only use this for full-screen elements that span the entire screen width. For smaller inline elements like this I suggest we don't use the `fullscreen` layout.
```suggestion
```

fischeversenker commented on src/components/checkbox/checkbox.stories.tsx:
```diff
@@ -0,0 +1,27 @@
+const Template: ComponentStory<typeof Checkbox> = (args) => (
+  <Checkbox {...args} />
+);
+export const Regular = Template.bind({});
```
Rather nitpicky, but important: Please add a new line between these two declarations.
```suggestion
);

export const Regular = Template.bind({});
```

fischeversenker commented on src/components/checkbox/checkbox.stories.tsx:
```diff
@@ -0,0 +1,27 @@
+      { name: 'white (figma)', value: '#F2F2F2;' },
+    ],
+  },
+};
```
Thanks for figuring this out. Is it specific to this component, though? Do we maybe want these to be our global storybook background options? I would say so, yes.
Feel free to add them to the exported `parameters` in `.storybook/preview.js` to make them global :raised_hands:

fischeversenker replied:
We agree that we are focusing on the light theme for now so there is no need to have these settings just yet. We will add them when we tackle the dark theme.

fischeversenker commented on src/components/checkbox/checkbox.tsx:
```diff
@@ -0,0 +1,85 @@
+  `;
+
+  const CheckboxContainer = styled.div`
+    display: flex;
```
`flex` turns this component into a "block element" which means that it will (almost) always be placed on a new line and will be followed by a new line. To make this more flexible we might want to turn this into `inline-flex` which would allow to have a checkbox "inline".
I'm not sure if this is an actual requirement but when using this component I would expect it to be an "inline element".

fischeversenker commented on src/components/checkbox/checkbox.tsx:
```diff
@@ -0,0 +1,85 @@
+    display: inline-block;
+  `;
+
+  const LabelText = styled(Text)`
```
The `Text` component has a `margin-bottom` of 16 pixels which does not quite fit here. I would also not recommend to nest a `p` element (`<p>` is a block element, `Text` is `styled.p`) inside a `label` element (`<span>` is an inline element). We should almost never nest block elements inside inline elements. Here a brief introduction into the "block"/"inline" topic: https://www.w3schools.com/html/html_blocks.asp

I suggest to create a new typography type for this special case that does not have any outer spacing as well as uses an inline element (`<span>` in this case) as its base.

fischeversenker commented on src/components/checkbox/checkbox.tsx:
```diff
@@ -0,0 +1,85 @@
+    display: inline-block;
+    fill: none;
+    margin-top: 5px;
+    margin-left: 5px;
```
These margins seem to push the checkbox to the bottom and thus break the alignment with the label:
![image](https://user-images.githubusercontent.com/7970458/141799417-76da9b4e-2d69-4a2a-acad-f54b5193a486.png)

Can you finetune this so the checkbox and the label are well aligned?

fischeversenker commented on src/components/checkbox/checkbox.tsx:
```diff
@@ -0,0 +1,85 @@
+
+  const handleChange = () => {
+    setChecked(!checked);
+    props.onChange(checked);
```
This will throw an exception if `props.onChange` is not defined. You can make this more failsafe by using ["optional chaining"](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Optional_chaining):
```suggestion
    props.onChange?.(checked);
```

fischeversenker reviewed (changes requested):
Thanks for this new component :sparkles:

It does not quite match the design specifications and there are a few improvements that I'd like to suggest. Please have a look at my comments.

georgiee commented:
Thank you for this PR and your efforts! And also thank you Felix for the review of this draft PR!

Forms & Form Elements are a little bit more complex. We also might need some common interface or such for them to work the same. Forms are only used in two places. I consider them a lower priority.

Maybe you feel better when you can deliver components more easily.  For example the quote component (https://github.com/satellytes/satellytes.com/issues/177). Might be more rewarding as I can see that this review could take a while to make it ready to use.

Any objections or would this be a good addition? In parallel this PR could be handled nonetheless but I fear it's going to take a while.

---

1. Besides this, I have also seen that you manually linked the related issue. You can simply type `#` and then search for the issue to link or put in the number of the issue. If it's in the text like `Related to # 123 (without the space)` somewhere in your text it will link automatically.
2. You seem to commit with great effort. I see good messages like `feat: introduce checkbox component` but I would like to point out that your PR is later ideally merged as a single commit (you can select that option). Which means you can happily commit multiple `chore: some information about what you did' or simply `chore: update` if you are tinkering around. If you are more comfortable with git you can also amend commits which is "updating the last commit". Ask us in the channel if you need some git assistance/knowledge 🙌

milljoniaer committed:
fix: right label alignment and smaller fixes
- align the label right
- align arrow right
- chaining to provide errors when onChange is not defined
- cleaning the DOM

Co-authored-by: Felix Hamann <fischeversenker@users.noreply.github.com>

fischeversenker commented on src/components/checkbox/checkbox.tsx:
```diff
@@ -0,0 +1,67 @@
+          />
+        </svg>
+      </StyledCheckbox>
+      <CheckboxLabelText>{props.label}</CheckboxLabelText>
```
If this component get's used without a label we still render this `CheckboxLabelText` element which leads to remaining whitespace. I suggest to render this element conditionally with something like the below code:
```suggestion
      {props.label && <CheckboxLabelText>{props.label}</CheckboxLabelText>}
```

fischeversenker commented on src/components/checkbox/checkbox.stories.tsx:
```diff
@@ -0,0 +1,17 @@
+
+export default {
+  component: Checkbox,
+  title: 'Components/Checkbox',
```
We can add an "[Action](https://storybook.js.org/docs/react/essentials/actions)" here to showcase the `onChange` property. With the below addition, Storybook will automatically print something in the Actions tab of the component by attaching a special Action callback to the `onChange` property.
```suggestion
  title: 'Components/Checkbox',
  argTypes: {
    onChange: { action: 'onChange' },
  },
```

fischeversenker reviewed (commented):
Thanks for the updates, @milljoniaer :pray:

I left two more comments. Please have a look before we merge this.

Also thanks a lot for raising your concerns, @georgiee.
I do agree that form elements are rather complex and it seems like with React this is not different. I just now read about [Controlled components](https://reactjs.org/docs/forms.html#controlled-components) but I don't see how this applies to a checkbox. @milljoniaer and I had a collaborative session where we took care of finetuning the layout and I now believe this is a solid checkbox component. What else would you expect from a component like this?

milljoniaer committed:
Merge branch 'main' into feat/checkbox

milljoniaer committed:
fix: add the suggested changes from the review
- render only if props.label is not falsy
- add Action for checkbox story
- fix returned boolean in onChange

milljoniaer commented:
I have added the requested changes by @fischeversenker. Please verify if they fit. 

Additionally I changed the code a little bit in, since I saw that onChanged returned the inverted boolean of checked in the Action. Is the react state-handling too slow?
But with the workaround in line 9 it works. Is that ok?

fischeversenker reviewed (approved):
> Additionally I changed the code a little bit in, since I saw that onChanged returned the inverted boolean of checked in the Action. Is the react state-handling too slow?

Yes, kind of. The react state-handling is "too slow" as in it's "asynchronous". So after the `setChecked()` function is done executing the state value is not yet updated. This has to do with React keeping track of all the changes that happen inside a component to know when to re-render the component. We will talk more about (a)synchronicity soon. For now, consider an asynchronous operation as something that happens in the background.

In my opinion the workaround that you came up with is fine. Another way of doing this would be to use the `useEffect()` hook which allows you to do something whenever the component's state changes. Take a look at the documentation about this "hook" [here](https://reactjs.org/docs/hooks-effect.html).

Since we will work on the `onChange` handler in the near future I suggest we keep your workaround and merge this as it is 🚀
